### PR TITLE
Make array ctor template on shape and stride parameters

### PR DESCRIPTION
**Pull request by SylvainCorlay**
I need to be able to initialize arrays with other sequence types than std::vector without copying strides and shape into a vector first.

### CODE REVIEW

File: include/pybind11/numpy.h [hunk 1]
```diff
@@ -362,21 +363,23 @@ class array : public buffer {
...
         : array(dt, shape, default_strides(shape, dt.itemsize()), ptr, base) { }
 
     array(const pybind11::dtype &dt, size_t count, const void *ptr = nullptr,
           handle base = handle())
         : array(dt, std::vector<size_t>{ count }, ptr, base) { }
 
-    template<typename T> array(const std::vector<size_t>& shape,
-                               const std::vector<size_t>& strides,
+    template<typename T, typename ST, typename DT>
+                         array(const ST& shape,
```
Comment (aldanor) [inline include/pybind11/numpy.h]: Minor: indentation seems off.

File: include/pybind11/numpy.h [hunk 2]
```diff
@@ -328,8 +328,9 @@ class array : public buffer {
 
     array() : array(0, static_cast<const double *>(nullptr)) {}
 
-    array(const pybind11::dtype &dt, const std::vector<size_t> &shape,
-          const std::vector<size_t> &strides, const void *ptr = nullptr,
+    template <typename ST, typename DT>
```
Comment (aldanor) [inline include/pybind11/numpy.h]: Maybe `s/ST/Shape/`, `s/DT/Strides/` (in other overloads too) so the signatures are a bit less cryptic?

**Comment by wjakob** [comment]
Merged, thanks!

**Comment by dean0x7d** [comment]
I think this PR was WIP. All the CI tests were failing with compiler errors.

**Comment by wjakob** [comment]
**REVERTED** -- I (mistakenly) assumed this was ready (it seemed like an innocuous change).

**Pull request merged**